src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs

```diff
@@ -65,6 +67,41 @@ private set
             }
         }
 
+        /// <summary>
+        /// Create InOutPorts on demand for the view
+        /// </summary>
+        public ItemsControl CreateInOutPortsControls(string name, int column, ObservableCollection<PortViewModel> itemsSource)
+        {
+            ItemsControl inoutportsControl = new ItemsControl();
+            inoutportsControl.Name = name;
+            inoutportsControl.ItemsSource = itemsSource;
+            inoutportsControl.Style = (Style)Resources["InOutPortControlStyle"];
+            Grid.SetRow(inoutportsControl, 2);
+            Grid.SetColumn(inoutportsControl, column);
+            Canvas.SetZIndex(inoutportsControl, 20);
+            inoutportsControl.HorizontalContentAlignment = HorizontalAlignment.Stretch;
+            grid.Children.Add(inoutportsControl);
+            return inoutportsControl;
+        }
+
+        private void PrepareInOutPortsControls()
+        {
+            if (ViewModel != null)
+            {
+                if (!inoutportsInitialized)
+                {
+                    inoutportsInitialized = true;
+                    CreateInOutPortsControls("inputPortControl", 0, ViewModel.InPorts);
+                    CreateInOutPortsControls("outputPortControl", 2, ViewModel.OutPorts);
+                }
+            }
+        }
+
+        private void NodeView_MouseEnter(object sender, MouseEventArgs e)
+        {
+            PrepareInOutPortsControls();
```

Comment:
what about destroying the ports control as well after zoom out or something to free up memory?

Reply:
that too yes